src/stubs-zlib.c

```diff
@@ -56,10 +57,21 @@ static void caml_zlib_error(char * fn, value vzs)
   mlraise(bucket);
 }
 
+void caml_zlib_free_stream(value vzs)
+{
+  caml_stat_free(ZStream_val(vzs));
+  ZStream_val(vzs) = NULL;
+}
+
+static struct custom_operations caml_zlib_stream_ops = {
+  "caml_zlib_stream_ops", &caml_zlib_free_stream, NULL, NULL, NULL, NULL
+};
+
 static value caml_zlib_new_stream(void)
 {
-  value res = alloc((sizeof(z_stream) + sizeof(value) - 1) / sizeof(value),
-                    Abstract_tag);
+  value res = caml_alloc_custom(&caml_zlib_stream_ops, sizeof(z_streamp), 0, 1);
+
+  ZStream_val(res) = caml_stat_alloc(sizeof(z_stream));
```

Comment:
+1 for getting rid of `alloc(..., Abstract_tag)`, which is old OCaml FFI style in at least two ways.